### ISSUE TITLE
Store created by dist.new_group doesn't appear to respect timeout

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#54524]
## 🐛 Bug

If default pg exists and user calls `dist.new_group` to create a new group, the corresponding prefix store created will have the default pg timeout as opposed to the timeout passed into `new_group`. It should probably have the timeout passed into `new_group` however.

This can result in issues if, say, we call `wait()` on a collective comm. issued by the new pg, and ProcessGroupNCCL calls into the store to broadcast unique nccl IDs, resulting in the wrong timeout being used. To reproduce, call `dist.new_group` with a small timeout, issue a collective comm. call that hangs, and log the timeout in FileStore.cpp (or whichever store implementation was used). 

Fix can probably be to set the right timeout here - https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L605

**Comment by SciPioneer** [comment]
Agree that we should not ignore the input timeout and silently use the timeout of the overall group.

Do we know any use case that need a timeout different from the overall process group? I guess only a short timeout for a subgroup makes sense.

**Comment by rohan-varma** [comment]
> Do we know any use case that need a timeout different from the overall process group? I guess only a short timeout for a subgroup makes sense.

My particular use case was for testing some debugability stuff, I wanted to create a pg that has a shorter timeout but that didn't work as expected since the store had the original, large timeout. There are possibly a few other cases where users want shorter timeouts for fail-fast/one-off comm. needs that they don't want to wait the entire original PG timeout for.

**Comment by SciPioneer** [comment]
> > Do we know any use case that need a timeout different from the overall process group? I guess only a short timeout for a subgroup makes sense.
> 
> My particular use case was for testing some debugability stuff, I wanted to create a pg that has a shorter timeout but that didn't work as expected since the store had the original, large timeout. There are possibly a few other cases where users want shorter timeouts for fail-fast/one-off comm. needs that they don't want to wait the entire original PG timeout for.

Indeed. Good for debugging mode which can fail fast. What about just setting a small timeout for the global group? Won't this work for your use case?

**Comment by rohan-varma** [comment]
Turns out that investigating this resulted in uncovering some further issues with TCPStore - 

1) PrefixStore sets the timer on the wrapped TCP (or other) store, so if we change this on a prefix store where the underlying store is shared it can unexpectedly affect other store timeouts. 

2) Timeout for TCPStore is not completely robust in that it is only implemented for `get` and not other things such as `set`. 

Due to (2) we can actually probably support this in the current way that things work by saving the set timeout on `PrefixStore`, and passing it into `get` call upon a get request. Although, this will ultimately make proper timeouts harder to implement down the line.
